GVFS/GVFS.Common/LibGit2RepoPool.cs

```diff
@@ -67,8 +81,27 @@ public bool TryInvoke<TResult>(Func<LibGit2Repo, TResult> function, out TResult
 
         private LibGit2Repo GetRepoFromPool()
         {
+            this.ResetTimer();
+
             LibGit2Repo repo;
-            if (this.pool.TryTake(out repo, TryTakeTimeoutMilliseconds))
+            if (this.pool.TryTake(out repo, TryTakeWaitShort))
+            {
+                return repo;
+            }
+
+            if (this.numDisposedRepos > 0)
+            {
+                if (Interlocked.Decrement(ref this.numDisposedRepos) >= 0)
+                {
+                    return this.repoFactory();
+                }
+                else

```

Comment:
Why do we increment the number of disposed repos here?

Reply:
If the `Interlocked.Decrement()` is below zero, it means that another thread decremented it to zero between the top `if` and the conclusion of the `Interlocked.Decrement()` statement. We need to increment the value back to ensure the value is correct (as we are not actually instantiating a new repo.

This is essentially what happens when our pool is _almost full_ and multiple requests for a repo happen at the same time.

Reply:
Ah - thanks for the explanation.

I think the edge condition logic is sufficiently subtle enough that the entire block could be encapsulated in its own method `TryCreateRepository(...)`, or at least a comment on the logic.

Did you consider using a [Semaphore](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphore?view=netframework-4.7.2) instead? I think this construct has the behavior you want, while avoiding some of need to re-increment on boundary conditions.

Unresolving this just to bring it back... If you decide to not address this, please resolve it again.

Reply:
Another possibility - just do the book keeping under a regular lock. I think it would be more clear than the special case handling of the edge condition...

Reply:
Semaphore has a [WaitOne(..)](https://docs.microsoft.com/en-us/dotnet/api/system.threading.waithandle.waitone?view=netframework-4.7.2#System_Threading_WaitHandle_WaitOne_System_Int32_System_Boolean_) with timeout.

We would just need to lock around the increment / decrement of the number of available repositories to allocate. 

i.e. 
1) take lock
2) test if we have available repositories to allocate, if so then decrement. 
3) release lock
4) create repo if we had available repositories to allocate.

I think these approaches could make the logic simpler - but I also think we could validate correctness with tests. Are we planning on including tests?

Reply:
I believe this use of `Interlocked` is a standard way of dealing with resources, and [interlocked is less resource-intensive than `lock()`](https://stackoverflow.com/a/154803/127088).

Tests are included in the latest push.